test/OpenTelemetry.Instrumentation.AspNetCore.Tests/RouteTests/RoutingTests.cs

```diff
@@ -0,0 +1,191 @@
+// <copyright file="RoutingTests.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#nullable enable
+
+using System.Diagnostics;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using OpenTelemetry;
+using OpenTelemetry.Metrics;
+using OpenTelemetry.Trace;
+using RouteTests.TestApplication;
+using Xunit;
+using static OpenTelemetry.Internal.HttpSemanticConventionHelper;
+
+namespace RouteTests;
+
+public class RoutingTests : IClassFixture<RoutingTestFixture>
+{
+    private const string OldHttpStatusCode = "http.status_code";
+    private const string OldHttpMethod = "http.method";
+    private const string HttpStatusCode = "http.response.status_code";
+    private const string HttpMethod = "http.request.method";
+    private const string HttpRoute = "http.route";
+
+    private readonly RoutingTestFixture fixture;
+    private readonly List<Activity> exportedActivities = new();
+    private readonly List<Metric> exportedMetrics = new();
+
+    public RoutingTests(RoutingTestFixture fixture)
+    {
+        this.fixture = fixture;
+    }
+
+    public static IEnumerable<object[]> TestData => RoutingTestCases.GetTestCases();
+
+    [Theory]
+    [MemberData(nameof(TestData))]
+    public async Task TestHttpRoute(RoutingTestCases.TestCase testCase, bool useLegacyConventions)
+    {
+        var configuration = new ConfigurationBuilder()
+            .AddInMemoryCollection(new Dictionary<string, string?> { [SemanticConventionOptInKeyName] = useLegacyConventions ? null : "http" })
+            .Build();
+
+        using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            .ConfigureServices(services => services.AddSingleton<IConfiguration>(configuration))
+            .AddAspNetCoreInstrumentation()
+            .AddInMemoryExporter(this.exportedActivities)
+            .Build()!;
+
+        using var meterProvider = Sdk.CreateMeterProviderBuilder()
+            .ConfigureServices(services => services.AddSingleton<IConfiguration>(configuration))
+            .AddAspNetCoreInstrumentation()
+            .AddInMemoryExporter(this.exportedMetrics)
+            .Build()!;
+
+        await this.fixture.MakeRequest(testCase.TestApplicationScenario, testCase.Path);
+
+        for (var i = 0; i < 10; i++)
```

Comment:
Is this just to wait for the activities to be exported? You could use `tracerProvider.ForceFlush()` instead?

Reply:
We introduced this pattern in our tests quite some time ago. If I recall it is because the client can receive the response prior to the server ending its Activity, so ForceFlush is not sufficient to prevent a flaky test.

Maybe we can find a better strategy, but for now I think this has proven to work ok for us.